infra_event_notifier/backends/datadog.py

```diff
@@ -0,0 +1,44 @@
+import time
+import json
+
+import urllib.request
+
+# from urllib.request import Request, urlopen
+from typing import Mapping
+
+
+def notify_datadog(
+    title: str,
+    text: str,
+    tags: Mapping[str, str],
+    datadog_api_key: str,
+    alert_type: str = "user_update",
+    epoch: int = None,
+) -> Mapping[str, str]:
+    """
+    Sends an event to Datadog.
+
+    :param title: Title of DD event
+    :param text: Body of event
+    :param tags: dict storing event tags
+    :param datadog_api_key: DD API key for sending events
+    :param alert_type: Type of event if using an event monitor, see https://docs.datadoghq.com/api/latest/events/
+    """
+    # API docs: https://docs.datadoghq.com/api/latest/events/#post-an-event
+    if not epoch:
+        epoch = int(time.time())
+    payload = {
+        "title": title,
+        "text": text,
+        "tags": [f"{k}:{v}" for k, v in tags.items()],
+        "date_happened": epoch,
+        "alert_type": alert_type,
+    }
+    jsonData = json.dumps(payload)
+    data = jsonData.encode("utf-8")
+    req = urllib.request.Request("https://api.datadoghq.com/api/v1/events", data=data)
+    req.add_header("DD-API-KEY", datadog_api_key)
+    req.add_header("Content-Type", "application/json; charset=utf-8")
+    with urllib.request.urlopen(req) as response:
+        res = json.loads(response.read().decode("utf-8"))
+        return res
```

Comment:
What do you really need to return ? You should be specific in what you want to give back to the callsite because then you are stuck in what they are going to use and unable to refactor. Promise as little as possible. 
You do not seem to be using the return value anyway, so just return None?

Reply:
Moreover the error management is leaking implementation details. 
When you send a request this way `urllib.request.urlopen(req)` you can get three responses:
1. an exception in case of a connection error
2. an HTTP200 with a valid response
3. any HTTP response that signals a failure (4xx, 5xx).
4. (technically you could get a 200 with an unexpected payload)

For the call site, which does not want to know about the datadog protocol (why would it use a datadog backend module otherwise ?), those should be coalesced into two:
1. A return value (or none) that makes sense for the library and that is abstract from the datadog payload. For example you could return the event id as that is the only thing that may be useful to the client. 
2. An exception. There is no reason for the call site (which does not care if we speak HTTP to datadog) to interpret HTTP status codes. 

So please interpret the return code and raise exception for all responses that do not tell you an event has been created.

Reply:
I'll remove returning the response, I don't think the DD API returns a status code - it either returns `{"status": "ok"}` or it raises an `HTTPError` automatically otherwise

example:
- good response
```
{'status': 'ok', 'event': {'id': 7645075181933775479, 'id_str': '7645075181933775479', 'title': 'test123', 'text': 'test', 'date_happened': 1719944856, 'handle': None, 'priority': None, 'related_event_id': None, 'tags': [], 'url': 'https://app.datadoghq.com/event/event?id=7645075181933775479'}}
```
- bad response (when sending `date_happened` as `-1`)
```
Traceback (most recent call last):
  File "infra_event_notifier/backends/datadog.py", line 45, in <module>
    send_event(
  File "infra_event_notifier/backends/datadog.py", line 39, in send_event
    with urllib.request.urlopen(req) as response:
  File "/Users/benmckerry/.pyenv/versions/3.8.18/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/Users/benmckerry/.pyenv/versions/3.8.18/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/Users/benmckerry/.pyenv/versions/3.8.18/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/Users/benmckerry/.pyenv/versions/3.8.18/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/Users/benmckerry/.pyenv/versions/3.8.18/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/Users/benmckerry/.pyenv/versions/3.8.18/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request
```